File: ananta_contracts/webrtc_security_negotiation.py
"""Canonical security negotiation contract used before WebRTC activation."""

from __future__ import annotations

import hashlib
import json
from dataclasses import asdict, dataclass
from typing import Mapping

SECURITY_MODES = ("transport_only", "opportunistic_e2ee", "strict_e2ee")
SECURITY_MODE_STRENGTH = {name: index for index, name in enumerate(SECURITY_MODES)}
ALLOWED_ALGORITHMS = frozenset({"AES-256-GCM", "ECDH-P256-HKDF-SHA256"})
ALLOWED_PAYLOAD_CLASSES = frozenset({"control", "media", "semantic", "bulk"})


class SecurityNegotiationError(ValueError):
    def __init__(self, reason_code: str) -> None:
        self.reason_code = reason_code
        super().__init__(reason_code)


@dataclass(frozen=True)
class SecurityProposal:
    version: int
    negotiation_id: str
    scope_kind: str
    scope_id: str
    sender_id: str
    recipient_id: str
    minimum_mode: str
    selected_mode: str
    algorithms: tuple[str, ...]
    key_epoch: int
    payload_classes: tuple[str, ...]
    expires_at_ms: int

    def canonical_dict(self) -> dict[str, object]:
        return asdict(self)


@dataclass(frozen=True)
class FinalSecurityContract:
    offer: SecurityProposal
    answer: SecurityProposal
    digest: str
    signature: str


def parse_security_proposal(raw: Mapping[str, object]) -> SecurityProposal:
    expected = {
        "version",
        "negotiation_id",
        "scope_kind",
        "scope_id",
        "sender_id",
        "recipient_id",
        "minimum_mode",
        "selected_mode",
        "algorithms",
        "key_epoch",
        "payload_classes",
        "expires_at_ms",
    }
    if set(raw) != expected:
        raise SecurityNegotiationError("negotiation_fields_invalid")
    if raw.get("version") != 1 or isinstance(raw.get("version"), bool):
        raise SecurityNegotiationError("negotiation_version_invalid")
    for field in ("negotiation_id", "scope_id", "sender_id", "recipient_id"):
        if not _id(raw.get(field)):
            raise SecurityNegotiationError("negotiation_identity_invalid")
    if raw.get("scope_kind") not in {"session", "room"}:
        raise SecurityNegotiationError("negotiation_scope_invalid")
    minimum = raw.get("minimum_mode")
    selected = raw.get("selected_mode")
    if minimum not in SECURITY_MODE_STRENGTH or selected not in SECURITY_MODE_STRENGTH:
        raise SecurityNegotiationError("security_mode_invalid")
    if SECURITY_MODE_STRENGTH[str(selected)] < SECURITY_MODE_STRENGTH[str(minimum)]:
        raise SecurityNegotiationError("security_mode_below_minimum")
    algorithms = _closed_string_tuple(raw.get("algorithms"), ALLOWED_ALGORITHMS, "algorithm_invalid")
    payload_classes = _closed_string_tuple(raw.get("payload_classes"), ALLOWED_PAYLOAD_CLASSES, "payload_class_invalid")
    epoch = raw.get("key_epoch")
    expiry = raw.get("expires_at_ms")
    if not isinstance(epoch, int) or isinstance(epoch, bool) or not 1 <= epoch <= 2**31 - 1:
        raise SecurityNegotiationError("epoch_invalid")
    if not isinstance(expiry, int) or isinstance(expiry, bool) or expiry < 1:
        raise SecurityNegotiationError("negotiation_expiry_invalid")
    return SecurityProposal(
        version=1,
        negotiation_id=str(raw["negotiation_id"]),
        scope_kind=str(raw["scope_kind"]),
        scope_id=str(raw["scope_id"]),
        sender_id=str(raw["sender_id"]),
        recipient_id=str(raw["recipient_id"]),
        minimum_mode=str(minimum),
        selected_mode=str(selected),
        algorithms=algorithms,
        key_epoch=epoch,
        payload_classes=payload_classes,
        expires_at_ms=expiry,
    )


def security_contract_digest(offer: SecurityProposal, answer: SecurityProposal) -> str:
    raw = json.dumps(
        {
            "domain": "ananta.webrtc.security-negotiation.v1",
            "offer": offer.canonical_dict(),
            "answer": answer.canonical_dict(),
        },
        sort_keys=True,
        separators=(",", ":"),
        allow_nan=False,
    ).encode()
    return hashlib.sha256(raw).hexdigest()


def _closed_string_tuple(value: object, allowed: frozenset[str], reason: str) -> tuple[str, ...]:
    if not isinstance(value, (list, tuple)) or not value or len(value) > len(allowed):
        raise SecurityNegotiationError(reason)
    if any(not isinstance(item, str) or item not in allowed for item in value):
        raise SecurityNegotiationError(reason)
    if len(set(value)) != len(value) or list(value) != sorted(value):
        raise SecurityNegotiationError("canonicalization_invalid")
    return tuple(value)


def _id(value: object) -> bool:
    return (
        isinstance(value, str)
        and 1 <= len(value) <= 128
        and value[0].isalnum()
        and all(char.isalnum() or char in "._:-" for char in value)
    )


__all__ = [
    "ALLOWED_ALGORITHMS",
    "ALLOWED_PAYLOAD_CLASSES",
    "FinalSecurityContract",
    "SECURITY_MODE_STRENGTH",
    "SecurityNegotiationError",
    "SecurityProposal",
    "parse_security_proposal",
    "security_contract_digest",
]

File: frontend-angular/src/app/services/webrtc-security-negotiation.ts
import { canonicalSecurityJson } from './webrtc-secure-envelope';

export interface StrictSecurityProposalV1 {
  version: 1;
  negotiation_id: string;
  scope_kind: 'session';
  scope_id: string;
  sender_id: string;
  recipient_id: string;
  minimum_mode: 'strict_e2ee';
  selected_mode: 'strict_e2ee';
  algorithms: ['AES-256-GCM', 'ECDH-P256-HKDF-SHA256'];
  key_epoch: number;
  payload_classes: ['bulk', 'control', 'semantic'];
  expires_at_ms: number;
}

export interface FinalStrictPairSecurityContractV1 {
  version: 1;
  negotiation_id: string;
  offer: StrictSecurityProposalV1;
  answer: StrictSecurityProposalV1;
  digest: string;
  signature: string;
  signature_algorithm: 'HMAC-SHA256';
}

export class SecurityNegotiationError extends Error {
  constructor(readonly reasonCode: string) { super(reasonCode); }
}

const ID_RE = /^[A-Za-z0-9][A-Za-z0-9._:-]{0,127}$/;
const DIGEST_RE = /^[a-f0-9]{64}$/;
const PROPOSAL_FIELDS = [
  'version', 'negotiation_id', 'scope_kind', 'scope_id', 'sender_id', 'recipient_id',
  'minimum_mode', 'selected_mode', 'algorithms', 'key_epoch', 'payload_classes',
  'expires_at_ms',
] as const;
const CONTRACT_FIELDS = [
  'version', 'negotiation_id', 'offer', 'answer', 'digest', 'signature',
  'signature_algorithm',
] as const;

/**
 * Validate the control-plane-finalized bilateral transcript and independently recompute
 * its digest before that digest is accepted by the signed peer-key package.
 */
export async function validateFinalStrictPairSecurityContract(
  raw: unknown,
  options: {
    scopeId: string;
    epoch: number;
    remoteMembershipId: string;
    localMembershipId?: string;
    nowMs?: number;
  },
): Promise<FinalStrictPairSecurityContractV1> {
  const value = closedObject(raw, CONTRACT_FIELDS, 'security_contract_fields_invalid');
  if (value['version'] !== 1 || value['signature_algorithm'] !== 'HMAC-SHA256') {
    throw new SecurityNegotiationError('security_contract_version_invalid');
  }
  if (!identifier(value['negotiation_id']) || !DIGEST_RE.test(String(value['digest'] ?? ''))
    || !DIGEST_RE.test(String(value['signature'] ?? ''))) {
    throw new SecurityNegotiationError('security_contract_authentication_invalid');
  }
  const offer = parseStrictProposal(value['offer'], options);
  const answer = parseStrictProposal(value['answer'], options);
  const membershipIds = new Set([offer.sender_id, offer.recipient_id]);
  if (
    offer.negotiation_id !== value['negotiation_id']
    || answer.negotiation_id !== value['negotiation_id']
    || offer.sender_id !== answer.recipient_id
    || offer.recipient_id !== answer.sender_id
    || offer.sender_id === offer.recipient_id
    || !membershipIds.has(options.remoteMembershipId)
    || (options.localMembershipId !== undefined && (
      options.localMembershipId === options.remoteMembershipId
      || membershipIds.size !== 2
      || !membershipIds.has(options.localMembershipId)
    ))
  ) {
    throw new SecurityNegotiationError('negotiation_binding_mismatch');
  }
  const digest = await sha256Hex(canonicalSecurityJson({
    domain: 'ananta.webrtc.security-negotiation.v1',
    offer,
    answer,
  }));
  if (digest !== value['digest']) throw new SecurityNegotiationError('security_contract_digest_mismatch');
  return {
    version: 1,
    negotiation_id: value['negotiation_id'] as string,
    offer,
    answer,
    digest,
    signature: value['signature'] as string,
    signature_algorithm: 'HMAC-SHA256',
  };
}

function parseStrictProposal(
  raw: unknown,
  options: { scopeId: string; epoch: number; nowMs?: number },
): StrictSecurityProposalV1 {
  const value = closedObject(raw, PROPOSAL_FIELDS, 'negotiation_fields_invalid');
  for (const field of ['negotiation_id', 'scope_id', 'sender_id', 'recipient_id'] as const) {
    if (!identifier(value[field])) throw new SecurityNegotiationError('negotiation_identity_invalid');
  }
  if (
    value['version'] !== 1
    || value['scope_kind'] !== 'session'
    || value['scope_id'] !== options.scopeId
    || value['minimum_mode'] !== 'strict_e2ee'
    || value['selected_mode'] !== 'strict_e2ee'
  ) throw new SecurityNegotiationError('security_downgrade_rejected');
  if (!Number.isSafeInteger(value['key_epoch']) || value['key_epoch'] !== options.epoch) {
    throw new SecurityNegotiationError('epoch_mismatch');
  }
  if (!Number.isSafeInteger(value['expires_at_ms']) || (value['expires_at_ms'] as number) <= (options.nowMs ?? Date.now())) {
    throw new SecurityNegotiationError('negotiation_expired');
  }
  if (!exactStringArray(value['algorithms'], ['AES-256-GCM', 'ECDH-P256-HKDF-SHA256'])) {
    throw new SecurityNegotiationError('algorithm_invalid');
  }
  if (!exactStringArray(value['payload_classes'], ['bulk', 'control', 'semantic'])) {
    throw new SecurityNegotiationError('payload_class_invalid');
  }
  return value as unknown as StrictSecurityProposalV1;
}

function closedObject(
  raw: unknown,
  expected: readonly string[],
  reasonCode: string,
): Record<string, unknown> {
  if (!raw || typeof raw !== 'object' || Array.isArray(raw)) {
    throw new SecurityNegotiationError(reasonCode);
  }
  const value = raw as Record<string, unknown>;
  if (Object.keys(value).length !== expected.length || expected.some((field) => !(field in value))) {
    throw new SecurityNegotiationError(reasonCode);
  }
  return value;
}

function identifier(value: unknown): value is string {
  return typeof value === 'string' && ID_RE.test(value);
}

function exactStringArray(value: unknown, expected: readonly string[]): boolean {
  return Array.isArray(value)
    && value.length === expected.length
    && value.every((item, index) => item === expected[index]);
}

async function sha256Hex(value: string): Promise<string> {
  const digest = await crypto.subtle.digest('SHA-256', new TextEncoder().encode(value));
  return Array.from(new Uint8Array(digest))
    .map((byte) => byte.toString(16).padStart(2, '0'))
    .join('');
}
